### PR TITLE
Daniela Di Iorio's Hotel

### PR DESCRIPTION
**Pull request by mystioreo**
# Hotel
Congratulations! You're submitting your assignment!

## Comprehension Questions
Question | Answer
:------------- | :-------------
Describe a design decision you had to make when working on this project. What options were you considering? What helped you make your final decision? | A big decision-making challenge I had was how to deal with Blocks of rooms.  My initial thought was that a Block should be a subclass of Hotel, because in a sense a Block is a miniature version of a hotel within the bigger Hotel (a block has its own special reservations and its own special room rate).  I started out this way and made Block a subclass of hotel.  A Hotel was meant to have instances of Block created within it.  I had trouble here because I couldn't figure out how a Hotel and a Block would share access to the same set of rooms.  I wanted to allow the Hotel to have full access to each Block's reservations, and to allow a Block to access the Hotel's reservations.  Eventually this got too complicated and I decided instead to have BlockReservation be a Reservation subclass.  This worked better for me because a Block really does behave like a special type of reservation.  I still needed to figure out how Reservations and BlockReservations would interact with each other within the hotel (specificially, if a room is in a block how does the Hotel then add a real reservation to it?).  I decided to add a different method for adding reservations to rooms in a block.  This creates more dependencies in my code (at a certain point the code actually checks the class of a Reservation object), so there may still be work that needs to be done here.
Describe a concept that you gained more clarity on as you worked on this assignment. | Because we did this assignment after reading POODR Ch.2, I thought more about making methods single-responsibility.  This project let me practice making lots of small helper methods in my Hotel class instead of writing really long and complex methods.  Some of my methods could still be broken down a bit more (which I would look at during refactoring).
Describe a _nominal test_ that you wrote for this assignment. | An example of a nominal test would be checking to see if a room can be booked given a valid date range and available room number.
Describe an _edge case test_ that you wrote for this assignment. |  An example of an edge case would be trying to book a reservation that has the same start date as another reservation's end date, or vice-versa.  This particular test helped me find a bug in my date-checking logic.
How do you feel you did in writing pseudocode first, then writing the tests and then the code? |  For the first two phases, I did a decent job getting my thoughts down on paper before starting to code.  I wrote the *names* of the tests before writing the code, but the test code mostly got written in parallel with the program code.  This was because I kept changing my mind about method names and arguments, and my tests needed to know this information.  For phase 3, I did not succeed in the pseudocode-->TDD technique.  I had trouble designing my room blocks abstractly, and found that I wanted to start writing code and testing its functionality with my existing code before settling on a design.  This was not very cost-effective because it resulted in me writing and deleting many lines of code (and their associated tests) and rolling back changes several times.

### CODE REVIEW

File: lib/reservation.rb [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+    @price_per_night = price_per_night
+    @price = calculate_price(checkin_date, checkout_date, price_per_night).to_i
+  end
+
+  def check_date_range(start, finish)
+    raise StandardError, 'Invalid date range' unless start < finish
+  end
+
+  def create_unique_id(room_number, checkin_date)
+    "#{room_number}000#{checkin_date.ld}".to_i
```
Comment (CheezItMan) [inline lib/reservation.rb:24]: How is this guaranteed to be unique?  It's likely to, but won't always be unique.

Also why have a method to get a Unique ID?  What purpose does it serve?

File: lib/reservation.rb [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+  end
+
+  def calculate_price(checkin_date, checkout_date, price_per_night)
+    (checkout_date - checkin_date) * price_per_night
```
Comment (CheezItMan) [inline lib/reservation.rb:28]: Just a note, that you have both the room and reservation tracking the price-per-night.  So some duplication there.

File: lib/reservation.rb [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+  end
+
+end
+
+class BlockReservation < Reservation
+  attr_reader :block_code, :room_list, :checkin_date, :checkout_date, :price, :discounted_rate
+  def initialize(checkin_date, checkout_date, room_list, discounted_rate, block_code)
```
Comment (CheezItMan) [inline lib/reservation.rb:35]: No method to reserve a room from within a block, or check to see if a room is in a block, or a lot of other things?  This class doesn't seem to add much over `Reservation`.

File: lib/hotel.rb [hunk 4]
```diff
@@ -0,0 +1,127 @@
+require 'pry'
+require_relative 'room'
+require_relative 'reservation'
+
+class Hotel
```
Comment (CheezItMan) [inline lib/hotel.rb:5]: One note on the name.  This class really doesn't do all the features of a Hotel.  Instead it manages Reservations.  So a more descriptive name would be appropriate.

File: lib/hotel.rb [hunk 5]
```diff
@@ -0,0 +1,127 @@
...
+    reservation = @reservations.find {|reservation| reservation.reservation_id == reservation_id}
+    return reservation.price
+  end
+
+  def find_available_rooms(start_date, end_date) #move some of this logic into Room class
+    available_room_numbers = []
+    room_list.each do |room|
+      room_is_available = true
+      (start_date..(end_date - 1)).each do |date|
+        if !room.is_available?(date)
```
Comment (CheezItMan) [inline lib/hotel.rb]: Does this check to see if a room is in a block?

File: lib/hotel_administrator.rb [hunk 6]
```diff
@@ -0,0 +1,88 @@
+#This is an extremely basic Command Line interface for the Hotel
+#There is no error checking or pretty formatting.
+#Mostly for experimentation purposes.
+
+#No need to grade this or provide feedback :)
```
Comment (CheezItMan) [inline lib/hotel_administrator.rb:5]: ok :)

File: spec/hotel_spec.rb [hunk 7]
```diff
@@ -0,0 +1,218 @@
...
+  it 'adds the reservation to the rooms reservation array' do
+    room_6 = @my_hotel.find_room_by_number(6)
+    expect(room_6.reservations.length).must_equal 1
+  end
+  it 'raises an ArgumentError if the room is not available for that date range' do
+    expect {
+      @my_hotel.make_reservation(Date.new(2010, 3, 7), Date.new(2010, 3, 8), 6, 'customer')
+    }.must_raise ArgumentError
+  end
+  it 'allows a new reservation to start on the last day of an existing reservation' do
```
Comment (CheezItMan) [inline spec/hotel_spec.rb:65]: 👍

File: spec/hotel_spec.rb [hunk 8]
```diff
@@ -0,0 +1,218 @@
...
+
+    date = Date.new(2010, 5, 5)
+    result = @my_hotel.find_reservations_by_date(date)
+    expect(result).must_be_instance_of Array
+    expect(result.length).must_equal 1
+
+    expect(result.all?{|item| item.is_a? Reservation}).must_equal true
+  end
+
+  it 'returns an empty array if no reservations exist on the given date' do
```
Comment (CheezItMan) [inline spec/hotel_spec.rb:95]: I would also test that it does not include reservations that **end** on that date.

File: spec/hotel_spec.rb [hunk 9]
```diff
@@ -0,0 +1,218 @@
...
+
+
+  it 'adds the reservation to each room and makes the rooms in the block unavailable during the given date range' do
+    (1..5).each do |num|
+      expect(@my_hotel.find_available_rooms(Date.new(2010, 3, 4), Date.new(2010, 5, 8))).wont_include num
+    end
+    expect(@my_hotel.find_available_rooms(Date.new(2010, 3, 4), Date.new(2010, 5, 8))).must_include 6
+  end
+
+  it 'prevents make_reservation from reserving a room in the block' do
```
Comment (CheezItMan) [inline spec/hotel_spec.rb:172]: 👍

**Comment by CheezItMan** [comment]
# Hotel
## What We're Looking For

Feature | Feedback
---     | ---
**Baseline** |
Used git regularly	| Reasonable number of commits and good commit messages
Answer comprehension questions	| Check, I'm glad you got some practice putting your thoughts down on paper first.
**Design** |
Each class is responsible for a single piece of the program | Check, well done
Classes are loosely coupled | Check, only `Hotel` is tightly coupled to other classes.
**Wave 1** |
List rooms | Check
Reserve a room for a given date range | Check
List reservations for a given date | Check
Calculate reservation price | Check
Invalid date range produces an error | Check
Test coverage | 100%
**Wave 2** |
View available rooms for a given date range | Check
Reserving a room that is not available produces an error | Check
**Wave 3** |
Create a block of rooms | Check
Check if a block has rooms | Check
Reserve a room from a block | Check
**Fundamentals** | 
Names variables, classes and modules appropriately | Check
Understanding of variable scope - local vs instance | Check
Can create complex logical structures utilizing variables | Check
Appropriately uses methods to break down tasks into smaller simpler tasks | Check
Understands the differences between class and instance methods | Check
Appropriately uses iterators and enumerables | Check
Appropriately writes and utilizes classes | Check, although your `Hotel` class is doing an awful lot of work.  I would see what functionality I could move to be the responsibility of `Block` or `Reservation`.
Appropriately utilizes modules as a mechanism to encapsulate functionality | No modules used!
**Wrap Up** |
There is a refactors.txt file | Check
The file provides a roadmap to future changes | Check
**Additional Feedback** | Very well done!  You hit all the learning goals.  You could do more at breaking code into separate methods and classes, but you have a solid design here.
